src/bin/server/transcoder/actor.rs

```diff
@@ -0,0 +1,215 @@
+use lru_disk_cache::LruDiskCache;
+use lru_disk_cache::ReadSeek;
+
+use server::temp::TemporaryFiles;
+use server::transcoder::errors;
+use server::transcoder::errors::ResultExt;
+use server::transcoder::TranscodeTarget;
+
+use std::collections::HashMap;
+use std::ffi::OsStr;
+use std::ffi::OsString;
+use std::io;
+use std::path::Path;
+use std::path::PathBuf;
+use std::process::Command;
+use std::process::Output;
+
+use actix::prelude::*;
+
+use futures::future;
+use futures::future::Shared;
+use futures::Future;
+
+use std::cell::RefCell;
+use std::rc::Rc;
+use tokio_process::CommandExt;
+
+pub struct Transcoder {
+    // Rc<RefCell<T>> combination is used to have multiple owning pointers to the same data with
+    // runtime borrow checking. Needed because satisfying the borrow checker is difficult (and less
+    // elegant) when working with futures.
+    /// Cache of all pre-transcoded items. Stored on disk. Persists across application restarts.
+    disk_cache: Rc<RefCell<LruDiskCache>>,
+
+    /// Cache of message keys to futures of things which are being converted now.
+    future_cache: Rc<RefCell<HashMap<OsString, Shared<ResponseFuture<(), io::Error>>>>>,
+    temp: TemporaryFiles,
+}
+
+impl Transcoder {
+    pub fn new(cache: LruDiskCache, temp: TemporaryFiles) -> Transcoder {
+        Transcoder {
+            disk_cache: Rc::new(RefCell::new(cache)),
+            future_cache: Rc::new(RefCell::new(HashMap::new())),
+            temp,
+        }
+    }
+
+    /// Transcodes the file requested by the message into a temporary path.
+    fn transcode<P: AsRef<Path>>(
+        &self,
+        msg: &TranscodeMessage<P>,
+    ) -> Box<Future<Item = (Output, PathBuf), Error = io::Error>> {
```

Comment:
I'm not sure whether it is possible to use less boxes.

Reply:
Use `impl Future<...>`